sublist/src/lib.rs

```diff
@@ -6,22 +12,181 @@ pub enum Comparison {
     Unequal,
 }
 
-pub fn sublist<T: PartialEq>(a: &[T], b: &[T]) -> Comparison {
+#[allow(dead_code)]
+pub fn is_superlist<T: PartialEq>(a: &[T], b: &[T]) -> bool {
+    a.windows(b.len()).any(|w| w == b)
+}
+
+#[allow(dead_code)]
+pub fn is_superlist_rayon<T: PartialEq + Sync>(a: &[T], b: &[T]) -> bool {
+    //! Parallelize the window comparisons using rayon.
+
+    a.par_windows(b.len()).any(|w| w == b)
+}
+
+#[allow(dead_code)]
+pub fn is_superlist_threads<T: PartialEq + Sync>(a: &[T], b: &[T]) -> bool {
+    //! Strateforwardly send each window comparison to a new thread (no pooling).
+
+    let a = Arc::new(a);
+    let b = Arc::new(b);
+
+    // scoped threads as in https://stackoverflow.com/a/32751956/9076659
+    thread::scope(|s| {
+        let (tx, rx) = mpsc::channel();
+        let mut thread_counter = 0;
+
+        for window in a.windows(b.len()) {
+            thread_counter += 1;
+            // create a new transmitter for each thread
+            let tx = tx.clone();
+
+            let b = b.clone();
+
+            s.spawn(move || {
+                // TODO: kill the thread if we already found a match
+                let are_equal = *b == window;
+                tx.send(are_equal).unwrap();
+            });
+        }
+
+        for _ in 0..thread_counter {
+            if rx.recv() == Ok(true) {
+                return true;
+            }
+        }
+
+        false
+    }) // no semicolon to return result
+}
+
+type Job = Box<dyn FnOnce() + Send + 'static>;
+
+struct Worker<'s> {
+    id: usize,
+    thread: thread::ScopedJoinHandle<'s, ()>,
+}
+
+impl<'s> Worker<'s> {
+    fn new(
+        id: usize,
+        receiver: Arc<Mutex<mpsc::Receiver<Job>>>,
+        scope: &'s thread::Scope<'s, '_>,
+    ) -> Worker<'s> {
+        let thread = scope.spawn(move || loop {
+            let job = receiver.lock().unwrap().recv().unwrap();
+
+            println!("Worker {} got a job; executing.", id);
+
+            job();
+        });
+
+        Worker { id, thread }
+    }
+}
+
+pub struct ScopedThreadPool<'s> {
+    max_threads: usize,
+    workers: Vec<Worker<'s>>,
+    sender: mpsc::Sender<Job>,
+}
+
+impl<'s> ScopedThreadPool<'s> {
+    /// Create a new ThreadPool.
+    ///
+    /// The size is the number of threads in the pool.
+    ///
+    /// # Panics
+    ///
+    /// The `new` function will panic if the size is zero.
+    pub fn new(size: usize, scope: &'s thread::Scope<'s, '_>) -> ScopedThreadPool<'s> {
+        assert!(size > 0);
+
+        let mut workers = Vec::with_capacity(size);
+        let (sender, receiver) = mpsc::channel();
+
+        let receiver = Arc::new(Mutex::new(receiver));
```

Comment:
This is unfortunate, it means that your threads have to synchronize on getting new input. Your use case really calls for a single-producer-multi-consumer channel, which the standard library does not provide.

There are several channel crates out there which do provide that functionality, I know of crossbeam, flume and kanal. [crossbeam](https://docs.rs/crossbeam/latest/crossbeam/channel/index.html) is probably the most mature and "standard" one, so I recommend going with that for a start. But it's good to know there are others.

Reply:
Now that I think about it, you wanted to use only the standard library for this part, right? In that case, maybe consider not using channels at all. One approach that would probably work is to wrap the iterator over windows in an `Arc<Mutex<_>>`. Haven't given it too much thought though.

Reply:
Yeah... I tried to abandon the threadpool idea at all and create `Arc<Mutex<VecDeque<Job>>>` in the last commit, but it did not work out also. Well, rayon somehow did it, and it has a lot of lifetime annotation in its classes, so maybe the last version (after [612fe9e](https://github.com/EvgeniiaVak/Rust-Exercism/pull/1/commits/612fe9e88dee21413c986bbcf2dfd9f89bc89052)) will be working if putting some annotations in the right places 🤔 so far the ones I've tried did not work though (I did not even push any).

Reply:
Ok, I managed to get it working by returning thread::scope and removing the Box around task functions.